DependencyScan: track library style

compnerd opened the pull request:
Track if the dependency is static or dynamic. This is in preparation for helping rename the static library to differentiate it from import libraries.

compnerd commented:
@swift-ci please test

cachemeifyoucan reviewed (commented):
Change is fine. Do you have any documentation or writing for what is the change you are planning to do?

I think it would be better if you do both in one PR since storing info into cache that is not currently used (and cache is also mostly not used) is not very useful.

compnerd commented:
I don't want to make that change as part of this change because its really a wide-spread change. The change is going to change the autolinked libraries in Swift. When statically linking Swift libraries on Windows, it will implicitly add a `lib` prefix to the library to differentiate it from the import library (e.g. `libArgumentParser.lib` vs `ArgumentParser.lib`). This is to help support @etcwilde's desire to remove `swift_static` someday.

compnerd committed:
AST,DependencyScan,IRGen,Serialization,Tooling: track library style

Track if the dependency is static or dynamic. This is in preparation for
helping rename the static library to differentiate it from import
libraries.

compnerd commented:
@swift-ci please test

cachemeifyoucan commented on lib/AST/ModuleDependencies.cpp:
```diff
@@ -516,19 +525,9 @@ swift::dependencies::registerCxxInteropLibraries(
-    }
+    if (Target.isOSDarwin() || Target.isOSLinux() || Target.isOSWindows())
+      RegistrationCallback(LinkLibrary{"swiftCxxStdlib", LibraryKind::Library,
+                                       hasStaticCxxStdlib});
```
This is missing a parameter.

compnerd replied:
The fourth parameter is optional, defaults to `false`.

cachemeifyoucan replied:
I see. I thought this parameter is for `forceLoad` and was wrongly used for static. Now I see what you mean by using static to select if the prefix has `lib` or not.

Will be good to have test for the change.

compnerd replied:
This is tested now - we had special cased the C++ interop libraries (which have test coverage) and will now be through this path. If there something else that you have in mind?

artemcm reviewed (commented):
For completeness, this change should introduce:
```
swiftscan_link_library_info_get_is_static
```
To `DependencyScan.h`.

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please smoke test

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test Windows platform

compnerd commented:
@swift-ci please test macOS platform

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test Windows platform

compnerd commented:
@swift-ci please test Windows platform

compnerd commented:
@swift-ci please test Windows platform

compnerd commented:
@swift-ci please smoke test macOS platform

compnerd commented:
@swift-ci please test Linux platform

compnerd commented:
@swift-ci please test macOS platform

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please smoke test macOS platform

compnerd commented:
@swift-ci please smoke test macOS platform

compnerd commented:
@swift-ci please smoke test Linux platform

compnerd commented:
@swift-ci please test Windows platform
